common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -979,81 +1093,127 @@ private boolean ensureThreadStarted(int oldState) {
     }
 
     private void doStartThread() {
-        assert thread == null;
         executor.execute(new Runnable() {
             @Override
             public void run() {
+                processingLock.lock();
+                assert thread == null;
                 thread = Thread.currentThread();
                 if (interrupted) {
                     thread.interrupt();
+                    interrupted = false;
                 }
-
                 boolean success = false;
                 updateLastExecutionTime();
+                boolean suspend = false;
                 try {
-                    SingleThreadEventExecutor.this.run();
-                    success = true;
+                    for (;;) {
+                        SingleThreadEventExecutor.this.run();
+                        success = true;
+
+                        int currentState = state;
+                        if (canSuspend(currentState)) {
+                            if (!STATE_UPDATER.compareAndSet(SingleThreadEventExecutor.this,
+                                    ST_SUSPENDING, ST_SUSPENDED)) {
+                                // Try again as the CAS failed.
+                                continue;
+                            }
+
+                            if (!canSuspend(ST_SUSPENDED)) {
+                                // Seems like there was something added to the task queue again in the meantime,
+                                // let's try to update the state to started and try again to execute the run loop.
+                                STATE_UPDATER.compareAndSet(SingleThreadEventExecutor.this,
+                                        currentState, ST_STARTED);
+                                continue;
+                            }
```

Comment:
We should only continue this thread if we successfully 'restart' the loop, otherwise it could mean that `startThread()` won the race and is now in the process of starting a new thread.
```suggestion
                            if (!canSuspend(ST_SUSPENDED) && STATE_UPDATER.compareAndSet(SingleThreadEventExecutor.this,
                                        ST_SUSPENDED, ST_STARTED)) {
                                // Seems like there was something added to the task queue again in the meantime but we
                                // were able to re-engage this thread as the event loop thread.
                                continue;
                            }
```

Reply:
I think both will work but yeah maybe lets do this to let this thread die as fast as possible if we already started a new one